packages/contracts-bedrock/test/libraries/TransientContext.t.sol

```diff
@@ -0,0 +1,220 @@
+
+    /// @notice Tests that `callDepth()` outputs the corrects call depth.
+    /// @param _callDepth Call depth to test.
+    function testFuzz_callDepth_succeeds(uint256 _callDepth) public {
+        assembly {
+            tstore(CALL_DEPTH_SLOT, _callDepth)
+        }
+        assertEq(TransientContext.callDepth(), _callDepth);
+    }
+
+    /// @notice Tests that `increment()` increments the call depth.
+    /// @param _startingCallDepth Starting call depth.
+    function testFuzz_increment_succeeds(uint256 _startingCallDepth) public {
+        vm.assume(_startingCallDepth < type(uint256).max);
+        assembly {
+            tstore(CALL_DEPTH_SLOT, _startingCallDepth)
+        }
+        assertEq(TransientContext.callDepth(), _startingCallDepth);
+
+        TransientContext.increment();
+        assertEq(TransientContext.callDepth(), _startingCallDepth + 1);
+    }
+
+    /// @notice Tests that `increment()` increments the call depth twice.
+    /// @param _startingCallDepth Starting call depth.
+    function testFuzz_increment_twice_succeeds(uint256 _startingCallDepth) public {
+        vm.assume(_startingCallDepth < type(uint256).max - 1);
+        testFuzz_increment_succeeds(_startingCallDepth);
+        assertEq(TransientContext.callDepth(), _startingCallDepth + 1);
+
+        TransientContext.increment();
+        assertEq(TransientContext.callDepth(), _startingCallDepth + 2);
+    }
+
+    /// @notice Tests that `decrement()` decrements the call depth.
+    /// @param _startingCallDepth Starting call depth.
+    function testFuzz_decrement_succeeds(uint256 _startingCallDepth) public {
+        vm.assume(_startingCallDepth > 0);
+        assembly {
+            tstore(CALL_DEPTH_SLOT, _startingCallDepth)
+        }
+        assertEq(TransientContext.callDepth(), _startingCallDepth);
+
+        TransientContext.decrement();
+        assertEq(TransientContext.callDepth(), _startingCallDepth - 1);
+    }
+
+    /// @notice Tests that `decrement()` decrements the call depth twice.
+    /// @param _startingCallDepth Starting call depth.
+    function testFuzz_decrement_twice_succeeds(uint256 _startingCallDepth) public {
+        vm.assume(_startingCallDepth > 1);
+        assembly {
+            tstore(CALL_DEPTH_SLOT, _startingCallDepth)
+        }
+        assertEq(TransientContext.callDepth(), _startingCallDepth);
+
+        TransientContext.decrement();
+        assertEq(TransientContext.callDepth(), _startingCallDepth - 1);
+
+        testFuzz_decrement_succeeds(_startingCallDepth - 1);
+    }
+
+    /// @notice Tests that `get()` returns the correct value.
+    /// @param _slot  Slot to test.
+    /// @param _value Value to test.
+    function testFuzz_get_succeeds(bytes32 _slot, uint256 _value) public {
+        assertEq(TransientContext.get(_slot), 0);
+
+        bytes32 tSlot = keccak256(abi.encodePacked(TransientContext.callDepth(), _slot));
```

Comment:
encode packed works here because its 2 bytes32 values but it wouldn't work if the types weren't 1 word each, its better to use `abi.encode` unless `encodePacked` is necessary

Reply:
In some ways this is a leaky abstraction - you are using `tstore` outside of the library to populate `get`. If the implementation under the hood changed, then this test would break. Ideally the tests are written in a way that doesn't leak the implementation details outside. This isn't always possible, like with the `tstore(CALL_DEPTH_SLOT, _startingCallDepth)` and `increment()` tests to ensure that the number goes up at various depths

Reply:
Ok i see the way that you use this test internally in the fuzz test below, i do like that pattern

Reply:
You can resolve this issue after you read it